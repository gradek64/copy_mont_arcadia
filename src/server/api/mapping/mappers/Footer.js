import Mapper from '../Mapper'

export default class Footer extends Mapper {
  mapEndpoint() {
    this.destinationEndpoint = '/webapp/wcs/stores/servlet/FooterDisplay'
    this.method = 'get'
  }
  mapRequestParameters() {
    const { catalogId, langId, siteId: storeId } = this.storeConfig
    this.query = {
      langId,
      storeId,
      catalogId,
    }
    this.payload = {}
  }
}
