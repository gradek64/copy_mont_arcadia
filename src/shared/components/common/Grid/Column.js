import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

@connect((state) => ({
  isResponsive: state.features.status.FEATURE_RESPONSIVE,
}))
class Column extends Component {
  static propTypes = {
    isResponsive: PropTypes.bool.isRequired,
    mobile: PropTypes.string,
    responsive: PropTypes.string,
    className: PropTypes.string,
    componentType: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  }

  static defaultProps = {
    children: null,
    mobile: 'col-12',
    componentType: 'div',
    className: '',
    responsive: 'col',
  }

  render() {
    const {
      children,
      isResponsive,
      mobile,
      responsive,
      className,
      componentType,
      dispatch,
      ...props
    } = this.props

    const ComponentType = componentType

    return (
      <ComponentType
        {...props}
        className={`${isResponsive ? responsive : mobile} ${className}`
          .trim()
          .replace(/\s+/g, ' ')}
      >
        {children}
      </ComponentType>
    )
  }
}

export default Column
