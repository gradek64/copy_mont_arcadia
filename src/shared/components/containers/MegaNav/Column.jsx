import React from 'react'
import PropTypes from 'prop-types'
import {
  isNil,
  lensPath,
  view,
  equals,
  type,
  map,
  filter,
  flatten,
  pipe,
  values,
} from 'ramda'
import classnames from 'classnames'

import Content from './Content'

import { heDecode } from '../../../lib/html-entities'

// Helpers
const noColumns = (category) =>
  isNil(category.columns) || category.columns.length <= 0

const lensImagePath = lensPath(['subcategories', 0, '', 0, 'image', 'span'])

const getColumnSpan = (column, lensImagePath) =>
  isNil(view(lensImagePath, column)) ? 1 : view(lensImagePath, column)

const getColumnWidth = (columnCount) => {
  const NUMBER_OF_COLUMNS = 4
  return `${(columnCount / NUMBER_OF_COLUMNS) * 100}%`
}

const isLastColumn = (category, column) => {
  if (category.columns) {
    return equals(Number(column.span), category.columns.length)
  }
}

const getColumnClassNames = (lastColumn) =>
  classnames('MegaNav-column', { 'MegaNav-lastColumn': lastColumn })

const renderSimplifiedList = (subcategories) => {
  return pipe(
    map((item) => values(item)),
    flatten,
    filter((subcategory) => !subcategory.isHidden),
    map((subcategory) => (
      <a
        key={`${subcategory.label}`}
        href={subcategory.redirectionUrl || subcategory.url}
      >
        {heDecode(subcategory.label)}
      </a>
    ))
  )(subcategories)
}

const Column = ({ category, column, hideMegaNav, isActive }) => {
  // NOTE:
  // column.subcategories is passed as two prop types Array and String
  //  - String is rendered as an empty column
  //  - Array should render subcategory links
  if (noColumns(category) || type(column.subcategories) !== 'Array') return null

  const columnSpan = getColumnSpan(column, lensImagePath)
  const lastColumn = isLastColumn(category, column)

  if (!isActive) return renderSimplifiedList(column.subcategories)

  return (
    <div
      className={getColumnClassNames(lastColumn)}
      style={{
        width: getColumnWidth(columnSpan),
      }}
    >
      {column.subcategories.map((block) => {
        return Object.keys(block).map((blockName, idx) => {
          const arraySubcategories = block[blockName]
          return (
            <Content
              // eslint-disable-next-line react/no-array-index-key
              key={`content_${idx}`}
              arraySubcategories={arraySubcategories}
              blockName={blockName}
              hideMegaNav={hideMegaNav}
            />
          )
        })
      })}
    </div>
  )
}

Column.propTypes = {
  category: PropTypes.shape({
    categoryId: PropTypes.string,
    columns: PropTypes.array,
    label: PropTypes.string,
    redirectionUrl: PropTypes.string,
    totalcolumns: PropTypes.number,
    url: PropTypes.string,
  }),
  column: PropTypes.shape({
    subcategories: PropTypes.array,
    span: PropTypes.string,
  }),
  hideMegaNav: PropTypes.func.isRequired,
  isActive: PropTypes.bool,
}

Column.defaultProps = {
  category: {},
  column: {
    subcategories: [],
  },
  isActive: true,
}

export default Column
