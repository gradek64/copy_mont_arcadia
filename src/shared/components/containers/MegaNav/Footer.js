/* MegaNav > Footer Component
 *
 * To render a footer container with the option to render:
 * - 1 column with the span of 4 (100%)
 * - 2 columns with the span of 2 each (50%)
 * - 4 columns with the span of 1 each (25%)
 *
 * NOTE:
 *  - The MegaNav is a four column layout.
 *  - Currently the "category" data can not be modified to include a footer property.
 *  - To identify if a footer image has been supplied a deep search inside category is required
 *       EXAMPLE: category > columns > subcategories > 0 > [name of property] > image > span === 'footer'
 *  */

import React from 'react'
import PropTypes from 'prop-types'
import { all, equals, dissoc, has, clone, forEachObjIndexed } from 'ramda'

// Components
import Column from './Column'

// Helpers
const hasFooterImage = (subcategory) => {
  const hasImage = has('image')
  const hasSpan = has('span')

  return (
    hasImage(subcategory) &&
    hasSpan(subcategory.image) &&
    subcategory.image.span === 'footer'
  )
}

const filterColumns = ({ columns }) =>
  columns.filter((column) => {
    const equalsFalse = equals(false)
    const subCategories = column.subcategories[0]
    const footerImage = []

    forEachObjIndexed(
      (value, key) =>
        subCategories[key].forEach((subCategory) =>
          footerImage.push(hasFooterImage(subCategory))
        ),
      subCategories
    )

    // If all values inside the array "footerImage" is false then remove subCategory
    return !all(equalsFalse, footerImage)
  })

const filterSubCategoriesWithoutFooter = (columns) =>
  columns.map((column) => {
    // Calculate the footer span
    // 4 divided by 1 column is a span of 4
    // 4 divided by 2 columns is a span of 2
    // 4 divided by 4 columns is a span of 1
    const span = 4 / columns.length
    const columnClone = clone(column)
    const subCategories = column.subcategories[0]

    forEachObjIndexed((value, key) => {
      const filteredSubCategories = subCategories[key].filter((t) =>
        hasFooterImage(t)
      )
      columnClone.subcategories[0] = dissoc(key, column.subcategories[0])
      columnClone.subcategories[0][''] = [...filteredSubCategories]
      columnClone.subcategories[0][''][0].image.span = span
    }, subCategories)
    return columnClone
  })

const removeSubCategoryWithoutFooter = (columns) =>
  columns.map((column) => {
    const columnClone = clone(column)
    const subCategories = column.subcategories[0]

    forEachObjIndexed((value, key) => {
      const filteredSubCategories = subCategories[key].filter((t) =>
        hasFooterImage(t)
      )

      if (filteredSubCategories.length === 0) {
        columnClone.subcategories[0] = dissoc(key, columnClone.subcategories[0])
      }
    }, subCategories)
    return columnClone
  })

const getFooterColumns = (category) => {
  const filteredColumns = filterColumns(category)
  const removedSubcategory =
    filteredColumns.length > 0
      ? removeSubCategoryWithoutFooter(filteredColumns)
      : []

  return removedSubcategory.length > 0
    ? filterSubCategoriesWithoutFooter(removedSubcategory)
    : []
}

const Footer = ({ category, hideMegaNav }) => {
  const footerColumns = getFooterColumns(category)
  if (footerColumns.length <= 0) return null
  return (
    <div className="MegaNav-subNav MegaNav-subNav--footer">
      {footerColumns.map((column, idx) => (
        <Column
          // eslint-disable-next-line react/no-array-index-key
          key={`column_${idx}`}
          category={category}
          column={column}
          hideMegaNav={hideMegaNav}
        />
      ))}
    </div>
  )
}

Footer.propTypes = {
  category: PropTypes.shape({
    categoryId: PropTypes.string,
    columns: PropTypes.array,
    label: PropTypes.string,
    redirectionUrl: PropTypes.string,
    totalcolumns: PropTypes.number,
    url: PropTypes.string,
  }),
  hideMegaNav: PropTypes.func.isRequired,
}

Footer.defaultProps = {
  category: {
    columns: [],
  },
}

export default Footer
