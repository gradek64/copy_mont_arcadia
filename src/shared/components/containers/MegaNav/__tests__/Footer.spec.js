import {
  buildComponentRender,
  shallowRender,
} from '../../../../../../test/unit/helpers/test-component'

import Footer from '../Footer'
import Column from '../Column'

const requiredProps = {
  category: {
    columns: [
      {
        subcategories: {
          0: {
            newIn: [
              { image: { span: 'footer' } },
              { label: 'something to buy' },
            ],
          },
        },
      },
      {
        subcategories: {
          0: {
            jeans: [{ image: { span: 'footer' } }, { label: 'blue jeans' }],
            0: [{ image: { span: 1 } }, { label: 'red jeans' }],
          },
        },
      },
      {
        subcategories: {
          0: {
            socks: [{ label: 'winter socks' }],
            jeans: [{ image: { url: '' } }],
          },
        },
      },
    ],
    url: 'http://www.topman.com',
  },
  hideMegaNav: () => {},
}

const filterColumnsProps = {
  category: {
    categoryId: '2244743',
    url: '/en/tsuk/category/brands-4210405',
    label: 'Brands',
    redirectionUrl: '',
    totalcolumns: 4,
    columns: [
      {
        span: '1',
        subcategories: [
          {
            'SHOP BY CATEGORY': [
              {
                categoryId: '2258006',
                url: '/en/tsuk/category/brands-4210405/beauty-brands-4233544',
                label: 'Beauty Brands',
                redirectionUrl: '',
              },
              {
                paddingTop: '',
                isHidden: false,
                bold: false,
                sale: false,
                image: {
                  width: 1110,
                  openNewWindow: true,
                  height: 140,
                  span: 'footer',
                  url:
                    '/wcsstore/ConsumerDirectStorefrontAssetStore/images/colors/color7/cms/pages/json/json-0000139861/images/BRANDS_LOGOS2.png',
                },
                categoryId: '3500058',
                url:
                  '/en/tsuk/category/brands-4210405/top-nav-shop-all-brands-7323864',
                label: 'Top Nav - Shop All Brands',
                redirectionUrl: '',
              },
            ],
          },
        ],
      },
      {
        span: '2',
        subcategories: [
          {
            '': [
              {
                categoryId: '3291092',
                url: '/en/tsuk/category/brands-4210405/ivy-park-5834347',
                label: 'Ivy Park',
                redirectionUrl:
                  '/en/tsuk/category/clothing-427/ivy-park-5463599?cat1=203984&cat2=3281520',
              },
              {
                categoryId: '3406848',
                url: '/en/tsuk/category/brands-4210405/adidas-6008354',
                label: 'adidas',
                redirectionUrl: '',
              },
            ],
          },
        ],
      },
    ],
  },
  hideMegaNav: () => {},
}

const filterColumnsExpected = {
  span: '1',
  subcategories: [
    {
      '': [
        {
          paddingTop: '',
          isHidden: false,
          bold: false,
          sale: false,
          image: {
            width: 1110,
            openNewWindow: true,
            height: 140,
            span: 4,
            url:
              '/wcsstore/ConsumerDirectStorefrontAssetStore/images/colors/color7/cms/pages/json/json-0000139861/images/BRANDS_LOGOS2.png',
          },
          categoryId: '3500058',
          url:
            '/en/tsuk/category/brands-4210405/top-nav-shop-all-brands-7323864',
          label: 'Top Nav - Shop All Brands',
          redirectionUrl: '',
        },
      ],
    },
  ],
}

describe('<Footer />', () => {
  const renderComponent = buildComponentRender(shallowRender, Footer)

  it('should render with two footer columns', () => {
    expect(renderComponent(requiredProps).getTree()).toMatchSnapshot()
  })

  it('returns nothing if has no columns passed', () => {
    const { wrapper } = renderComponent({
      ...requiredProps,
      category: { columns: [] },
    })
    expect(wrapper.find(Column)).toHaveLength(0)
  })

  it('returns Columns with correctly filtered columns passed to Footer', () => {
    const { wrapper } = renderComponent(filterColumnsProps)
    expect(wrapper.find(Column).get(0).props.column).toMatchObject(
      filterColumnsExpected
    )
  })
})
