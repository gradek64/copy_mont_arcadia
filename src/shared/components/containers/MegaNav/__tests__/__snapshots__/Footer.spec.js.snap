// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer /> should render with two footer columns 1`] = `
<div
  className="MegaNav-subNav MegaNav-subNav--footer"
>
  <Column
    category={
      Object {
        "columns": Array [
          Object {
            "subcategories": Object {
              "0": Object {
                "newIn": Array [
                  Object {
                    "image": Object {
                      "span": "footer",
                    },
                  },
                  Object {
                    "label": "something to buy",
                  },
                ],
              },
            },
          },
          Object {
            "subcategories": Object {
              "0": Object {
                "0": Array [
                  Object {
                    "image": Object {
                      "span": 1,
                    },
                  },
                  Object {
                    "label": "red jeans",
                  },
                ],
                "jeans": Array [
                  Object {
                    "image": Object {
                      "span": "footer",
                    },
                  },
                  Object {
                    "label": "blue jeans",
                  },
                ],
              },
            },
          },
          Object {
            "subcategories": Object {
              "0": Object {
                "jeans": Array [
                  Object {
                    "image": Object {
                      "url": "",
                    },
                  },
                ],
                "socks": Array [
                  Object {
                    "label": "winter socks",
                  },
                ],
              },
            },
          },
        ],
        "url": "http://www.topman.com",
      }
    }
    column={
      Object {
        "subcategories": Object {
          "0": Object {
            "": Array [
              Object {
                "image": Object {
                  "span": 2,
                },
              },
            ],
          },
        },
      }
    }
    hideMegaNav={[Function]}
    isActive={true}
    key="column_0"
  />
  <Column
    category={
      Object {
        "columns": Array [
          Object {
            "subcategories": Object {
              "0": Object {
                "newIn": Array [
                  Object {
                    "image": Object {
                      "span": "footer",
                    },
                  },
                  Object {
                    "label": "something to buy",
                  },
                ],
              },
            },
          },
          Object {
            "subcategories": Object {
              "0": Object {
                "0": Array [
                  Object {
                    "image": Object {
                      "span": 1,
                    },
                  },
                  Object {
                    "label": "red jeans",
                  },
                ],
                "jeans": Array [
                  Object {
                    "image": Object {
                      "span": "footer",
                    },
                  },
                  Object {
                    "label": "blue jeans",
                  },
                ],
              },
            },
          },
          Object {
            "subcategories": Object {
              "0": Object {
                "jeans": Array [
                  Object {
                    "image": Object {
                      "url": "",
                    },
                  },
                ],
                "socks": Array [
                  Object {
                    "label": "winter socks",
                  },
                ],
              },
            },
          },
        ],
        "url": "http://www.topman.com",
      }
    }
    column={
      Object {
        "subcategories": Object {
          "0": Object {
            "": Array [
              Object {
                "image": Object {
                  "span": 2,
                },
              },
            ],
          },
        },
      }
    }
    hideMegaNav={[Function]}
    isActive={true}
    key="column_1"
  />
</div>
`;
